go-controller/pkg/ovn/obj_retry.go

```diff
@@ -126,18 +189,1067 @@ func (r *retryObjs) getObjRetryEntry(key string) *retryObjEntry {
+						objectsToRetry.oType, err)
+					return
+				}
+
+				// skip the whole update if the new object doesn't exist anymore in the API server
+				newer, err = oc.getResourceFromInformerCache(objectsToRetry.oType, newKey)
+				if err != nil && kerrors.IsNotFound(err) {
+					// QUESTION if not found in the API server, shouldn't we actually delete it?
+					// (this part is from the pods retry code)
+					klog.Infof("%v %s not found in the informers cache,"+
+						" not going to update it", objectsToRetry.oType, newKey)
+					return
+				}
+
+				klog.Infof("Update event received for resource %v, oldKey=%s, newKey=%s",
+					objectsToRetry.oType, oldKey, newKey)
+
+				objectsToRetry.skipRetryObj(newKey)
+				hasUpdateFunc := hasResourceAnUpdateFunc(objectsToRetry.oType)
+
+				// STEP 1:
+				// Delete existing (old) object if:
+				// a) it has a retry entry marked for deletion or
+				// b) the resource is in terminal state (e.g. pod is completed) or
+				// c) this resource type has no update function, so an update means delete old obj and add new one
+				retryEntry := objectsToRetry.getObjRetryEntry(oldKey)
+				if retryEntry != nil && retryEntry.oldObj != nil {
+					// [step 1a] there is a retry entry marked for deletion
+					klog.Infof("Found old retry object for %v %s: will delete it",
+						objectsToRetry.oType, oldKey)
+					if err := oc.deleteResource(objectsToRetry, retryEntry.oldObj,
+						retryEntry.config); err != nil {
+						klog.Errorf("Failed to delete stale object %s, during update: %v", oldKey, err)
+						oc.recordErrorEvent(objectsToRetry.oType, retryEntry.oldObj, err)
+						objectsToRetry.initRetryObjWithAdd(newer, newKey)
+						objectsToRetry.unSkipRetryObj(oldKey)
+						return
+					}
+					// remove the old object from retry entry since it was correctly deleted
+					objectsToRetry.removeDeleteFromRetryObj(oldKey)
+
+				} else if oc.isObjectInTerminalState(objectsToRetry.oType, newer) { // check the latest status on newer
+					// [step 1b] The object is in a terminal state: delete it from the cluster,
+					// delete its retry entry and return. This only applies to pod watchers
+					// (pods + dynamic network policy handlers watching pods).
+					oc.processObjectInTerminalState(objectsToRetry, old, oldKey, "update")
+					return
+
+				} else if !hasUpdateFunc {
+					// [step 1c] if this resource type has no update function,
+					// delete old obj and then add the new one
+					var existingCacheEntry interface{}
+					if retryEntry != nil {
+						existingCacheEntry = retryEntry.config
+					}
+					klog.Infof("Deleting old %s of type %s during update", oldKey, objectsToRetry.oType)
+					if err := oc.deleteResource(objectsToRetry, old, existingCacheEntry); err != nil {
+						klog.Errorf("Failed to delete %s %s, during update: %v",
+							objectsToRetry.oType, oldKey, err)
+						oc.recordErrorEvent(objectsToRetry.oType, old, err)
```

Comment:
as mentioned to u some resources update status and metrics for example egfw so how can this be handled

Reply:
My code here can be easily adapted to manage egress firewalls as well. The metrics part will go to `recordAddEvent` and `recordDeleteEvent`;  I will have to add a `recordUpdateEvent`, which wasn't necessary in the case of pods. 

As for the updating the status of the CRD, we can have it in `WatchResource` and it will get executed only for our CRDs (egressfirewalls, egressIPs, CloudPrivateIPConfig... I think that should be it). Either that or we put it inside the resource-specific part of `addResource`, `updateResource`, `deleteResource`.

I can take care of this once this PR is merged :)